Add support for writing AIDL FMQs

mlehtima opened the pull request.

mlehtima commented:
Use in https://github.com/sailfishos/sensorfw/pull/31

monich commented:
from the first glance it seems more appropriate to implement this as a protocol specific feature via yet another [`GBinderRpcProtocol`](https://github.com/mer-hybris/libgbinder/blob/master/src/gbinder_rpc_protocol.h) callback.

I generally don't like _"virtialuzation"_ by means of ifs, and on top of that, relying on protocol naming conventions is imo too fragile, especially when those conventions are not clearly spelled out and not guaranteed to be followed.

let me think...

monich commented on src/gbinder_rpc_protocol.c:
```diff
@@ -33,6 +33,7 @@
 #include "gbinder_rpc_protocol.h"
 #include "gbinder_reader.h"
 #include "gbinder_writer.h"
```
`gbinder_writer_p.h` includes `gbinder_writer.h` (which is generally the case for private headers with public counterparts), it doesn't make sense to include both

mlehtima replied:
Removed the include after changing code to use writer functions instead of writer_data functions.

monich commented on src/gbinder_writer.c:
```diff
@@ -1282,19 +1344,73 @@ gbinder_writer_data_append_fmq_descriptor(
     GBinderWriter* self,
     const GBinderFmq* queue) /* since 1.1.14 */
 {
+#if GBINDER_FMQ_SUPPORTED
```
alternatively, `gbinder_writer_append_fmq_descriptor` may return a boolean indicating whether fmq's are supported or not. that way the application can react on that at run time

monich replied:
and it's fine to add the return value, that won't break the abi

mlehtima replied:
Changed this back to how it was.

monich commented on src/gbinder_writer_p.h:
```diff
@@ -165,6 +165,20 @@ gbinder_writer_data_append_local_object(
+    const GBinderFmq* queue);
+
+#endif /* GBINDER_FMQ_SUPPORTED */
+
```
how about moving these implementations inside `gbinder_rpc_protocol_hidl_write_fmq_descriptor` and `gbinder_rpc_protocol_aidl_write_fmq_descriptor`? that would significantly reduce the number of `#if GBINDER_FMQ_SUPPORTED` blocks and generally makes sense imo (considering that these functions are only called from those callbacks)

mlehtima replied:
Moved the implementations to gbinder_rpc_protocol.c.

monich commented on src/gbinder_rpc_protocol.h:
```diff
@@ -34,6 +34,7 @@
 #define GBINDER_RPC_PROTOCOL_H
 
 #include "gbinder_types_p.h"
+#include "gbinder_fmq_p.h"
```
there's no need to include it from `gbinder_rpc_protocol.h` , `GBinderFmq` is defined in `gbinder_types.h`  and it's enough to compile the definition of the `write_fmq_descriptor` callback.

monich replied:
ah, ok. one more reason to make `write_fmq_descriptor` unconditional :)

mlehtima replied:
Include removed.

mlehtima commented:
Also added missing implementation for gbinder_writer_append_fds and updated the version info for that because it was never functional in the old version.

mlehtima commented on include/gbinder_writer.h:
```diff
@@ -257,6 +258,16 @@ gbinder_writer_append_parcelable(
+gbinder_writer_append_parcelable_finish(
+    GBinderWriter* writer,
+    gssize offset); /* Since 1.1.43 */
+
```
Those can be useful when writing unknown size parcelables but of course the same thing can be done manually. I can move those to gbinder_writer_p.h if that is preferred

mlehtima replied:
Moved to gbinder_writer_p.h now.

monich commented on src/gbinder_rpc_protocol.c:
```diff
@@ -78,6 +82,61 @@ static const GBinderRpcProtocol DEFAULT_PROTOCOL;
+ * Common AIDL protocol
+ *==========================================================================*/
+
+void
```
and I guess in order to compile on a platform where fmqs are not supported, it should be defined like this:
```c
#if GBINDER_FMQ_SUPPORTED
static
void
gbinder_rpc_protocol_aidl_write_fmq_descriptor(
    GBinderWriter* writer, const GBinderFmq* queue)
{
...
}
#else
#  define gbinder_rpc_protocol_aidl_write_fmq_descriptor NULL
#endif /* GBINDER_FMQ_SUPPORTED */

```

mlehtima replied:
Done
